dex/networks/erc20/contracts/ERC20SwapV0.sol

```diff
@@ -0,0 +1,207 @@
+    {
+        return swaps[secretHash];
+    }
+
+    // Initiation is used to specify the information needed to initiatite a swap.
+    // The ERC20 token address is provided separately to the initiate function.
+    struct Initiation {
+        uint refundTimestamp;
+        bytes32 secretHash;
+        address participant;
+        uint value;
+    }
+
+    // initiate initiates an array of swaps all for the same ERC20 token.
+    // It checks that all of the swaps have a non zero redemptionTimestamp
+    // and value, and that none of the secret hashes have ever been used
+    // previously. Once initiated, each swap's state is set to Filled. The
+    // tokens equal to the sum of each swap's value are now in the custody of
+    // the contract and can only be retrieved through redeem or refund.
+    //
+    // It is important to note that this calls an external contract which has no
+    // restrictions on gas used. This has the potential to open the contract up
+    // to a reentry attack.
+    //
+    // This is a writing function and requires gas. Failure or success should
+    // be guaged by querying the swap and checking state after being mined. Gas
+    // is expended either way.
+    function initiate(Initiation[] calldata initiations, address token)
+        public
+        senderIsOrigin()
+    {
+        uint initVal = 0;
+        for (uint i = 0; i < initiations.length; i++) {
+            Initiation calldata initiation = initiations[i];
+            Swap storage swapToUpdate = swaps[initiation.secretHash];
+
+            require(initiation.value > 0);
+            require(initiation.refundTimestamp > 0);
+            require(swapToUpdate.state == State.Empty);
+
+            swapToUpdate.initBlockNumber = block.number;
+            swapToUpdate.refundBlockTimestamp = initiation.refundTimestamp;
+            swapToUpdate.initiator = msg.sender;
+            swapToUpdate.participant = initiation.participant;
+            swapToUpdate.value = initiation.value;
+            swapToUpdate.state = State.Filled;
+            swapToUpdate.token = token;
+
+            initVal += initiation.value;
+        }
+		bool success;
+		bytes memory data;
+        (success, data) = token.call(abi.encodeWithSelector(TRANSFER_FROM_SELECTOR, msg.sender, address(this), initVal));
+        require(success && (data.length == 0 || abi.decode(data, (bool))), 'transfer from failed');
+    }
+
+    // Redemption is used to specify the information needed to redeem a swap.
+    struct Redemption {
+        bytes32 secret;
+        bytes32 secretHash;
+    }
+
+    // isRedeemable returns whether or not a swap identified by secretHash
+    // can be redeemed using secret.
+    function isRedeemable(bytes32 secretHash, bytes32 secret)
+        public
+        view
+        returns (bool)
+    {
+        Swap storage swapToRedeem = swaps[secretHash];
+        return swapToRedeem.state == State.Filled &&
+               swapToRedeem.participant == msg.sender &&
```

Comment:
Probably late to the party, but the `swapToRedeem.participant == msg.sender` would mean that this can only ever be checked by one wallet, right? I don't know if we wanted this to be observable from web browsers and stuff. It would just return false if not asked from the participant. No need to change here though, if it was not intended, because it's also like this in the eth version.

Reply:
No, any address can be used. When calling a `view` function a signature is required, so any address can be put into the `From` field.

Reply:
You mean a signature is **not** required?  No transaction or gas either, just an `accounts/abi/bind.(*BoundContract).Call` on the contract method with `&bind.CallOpts{From: whatever}`.

It's conceivable that this function could be written like `function isRedeemable(bytes32 secretHash, bytes32 secret, acct address) public view returns (bool)` so you could do it on say etherscan from the `#readContract` part of the page (e.g. https://etherscan.io/address/0xdac17f958d2ee523a2206206994597c13d831ec7#readContract).  However, since you would **never** want to execute this function outside of your own machine, except when actually redeeming, because of the `secret` argument, it need not be like that.  Also, we have the `swap(bytes32 secretHash) public view returns(Swap memory)` that presumably would be usable for free browser-based queries.

Reply:
> When calling a view function a signature is required

Not required? Interesting.

> since you would never want to execute this function outside of your own machine

oh yeah, derp, wouldnt want to expose the secret

What was it then we wanted to be able to do from explorers, or am I making stuff up. Just view swap state?